### ISSUE TITLE
External ecr-credential-provider not registered on kubelet in v20230322

### ISSUE THREAD

**Issue by GDegrove** [awslabs/amazon-eks-ami#1239]
**What happened**:
Latest AMI version build compatibility to the `--image-credential-provider-config`, in the file `install-worker.sh` the file is sed based on the presence of a specific tag (`--cloud-provider aws`)

This line was changed, and the current install-worker.sh is not able to enable that plugin anymore. (or configure kubelet)

**What you expected to happen**:

We expected that plugin to be enabled by default, or to be notified that this was changed and find a solution to resolve this.

**How to reproduce it (as minimally and precisely as possible)**:
Install latest version of EKS AMI, the  configuration injected at this line are missing: 
https://github.com/awslabs/amazon-eks-ami/blob/v20230322/scripts/install-worker.sh#L167

**Anything else we need to know?**:

**Environment**:
- AWS Region: eu-west-1
- Instance Type(s): AMD64
- EKS Platform version (use `aws eks describe-cluster --name <name> --query cluster.platformVersion`): 
- Kubernetes version (use `aws eks describe-cluster --name <name> --query cluster.version`): 1.22+ version are impacted
- AMI Version: AMI Release v20230322 
- Kernel (e.g. `uname -a`): 5.10
- Release information (run `cat /etc/eks/release` on a node):

```
```

**Comment by cartermckinnon** [comment]
One way to mitigate this would be to pass these flags with `--kubelet-extra-args`/`KUBELET_EXTRA_ARGS` during node bootstrap:
```
--image-credential-provider-config /etc/eks/ecr-credential-provider/ecr-credential-provider-config --image-credential-provider-bin-dir /etc/eks/ecr-credential-provider
```

Or, if you use a custom launch template but not a custom AMI, you can do something like this prior to the bootstrap:
```
  IMAGE_CREDENTIAL_PROVIDER_FLAGS='\\\n    --image-credential-provider-config /etc/eks/ecr-credential-provider/ecr-credential-provider-config \\\n   --image-credential-provider-bin-dir /etc/eks/ecr-credential-provider'
  sudo sed -i s,"KUBELET_EXTRA_ARGS","KUBELET_EXTRA_ARGS $IMAGE_CREDENTIAL_PROVIDER_FLAGS", /etc/eks/containerd/kubelet-containerd.service
```

I'll put up a fix momentarily. Apologies for the hassle, the existing `sed` is very fragile.

**Comment by cartermckinnon** [comment]
Reopening until an AMI release is available with the fix in #1240.

**Comment by cartermckinnon** [comment]
Sorry for the confusion -- I closed this issue inadvertently. A fix has been merged here, but the ensuing AMI release is not yet available. It's currently working its way through our release pipeline, and we expect it to be available in all regions within 24 hours.

**Comment by GDegrove** [comment]
Thanks, 

no worries, we will keep a close 👁️ on this. 
Thanks again for the swift feedback. 

Is there any plan to decommission this in the future? Or are we safe for some time?

**Comment by dc2tom** [comment]
Hi, with the workaround https://github.com/awslabs/amazon-eks-ami/issues/1239#issuecomment-1490658274 included I can see that the kubelet flags are now set and the credentials provider binary and configuration are in the correct place however when I hop onto a node in my cluster and run "ctr image pull xxxx.dkr.ecr.etc.etc" for a private image in another AWS account it will not allow me to pull.  Containerd will only authenticate if the repository is located within the same account.  

All the required IAM is in place, if I install docker on the node and run a login I can pull the images cross account.  Is there something I am missing that we need to configure to allow containerd to utilise the credentials-provider?

Thanks

**Comment by GDegrove** [comment]
Hello @cartermckinnon, 

Do you have an ETA on the release of the new AMI? 

Thanks,

Gil

**Comment by dc2tom** [comment]
At the present time (eu-west-1, using Karpenter) we are having to use the userdata workaround detailed above

**Comment by cartermckinnon** [comment]
Sorry for the delay -- given the limited impact of this change, and the possible mitigations, we decided to roll this fix out along with our initial 1.26 release. That build is working its way through our release process now, and will be available in all regions within the next few days.
